### PR TITLE
[forge] fix latency experiment with netem caveats

### PR DESCRIPTION
**Pull request by rustielin**
### Description

This PR overhauls the way we're applying network emulation chaos (delay and loss NetworkChaos in ChaosMesh terminology) such that the desired outcome is applied. Using the new chaos primitives, I've written a three-region network simulation between us-west, af-south, and eu-north, some of the highest latency DC pairs. This PR is meant to be a starting point for network chaos tests. We can tune the SuccessCriteria and chaos accordingly when iterating.

#### Three Region Simulation

Using https://www.cloudping.co/grid as a baseline, split the validators in the test into three equally sized regions. There is latency injected pairwise between each region-pair (symmetrically), and no latency between the validators and the transaction-emitter. Bandwidth is also limited across the entire namespace to 100mbps.

delay NetworkChaos is defined as:

```
# us-west <--> af-south
            latency_ms: 300,
            jitter_ms: 50,
            correlation_percentage: 50,

# us-west <--> eu-north
            latency_ms: 150,
            jitter_ms: 50,
            correlation_percentage: 50,

# eu-north <--> af-south
            latency_ms: 200,
            jitter_ms: 50,
            correlation_percentage: 50,
```

#### Update ChaosMesh templates to get the desired outcome

To correctly apply latency to Forge experiments, some deeper understanding of Chaos Mesh is required. 

Major learnings from: https://github.com/chaos-mesh/chaos-mesh/issues/3302
* Forge hits k8s Services rather than k8s PodIPs directly, so chaos is bypassed when using a pod Selector. Instead, use label selectors and change the service type to `NodePort` to get the desired effect
* The `from` direction is the only one that will work since NetworkChaos with types `delay` (latency) and `loss` due to the fact that it uses `netem` under the hood: 

> However, netem chaos (delay, loss, ...) will not work, because natively, the netem only affect outbound traffic, so that we only limit the outbound on source.

Also, currently only one type of netem chaos is allowed at a time. I've filed an issue in ChaosMesh https://github.com/chaos-mesh/chaos-mesh/issues/3631. For now, this means we cannot apply `delay` and `loss` at the same time, so we default to just doing `delay`. Bandwidth and partition seem to be unaffected by this, so in the three-region simulation, we're also limiting by 100mbps

### Test Plan

To test the chaos part, test on rustienet a number of scenarios manually. Use the test results as well as some common utilities to verify chaos is applied as intended (e.g. exec into pods and curl all other pods a bunch of times, taking the average latency)

1. 2 validators, 1 test runner all in the same namespace. This affects the test runner's performance
3. 2 valdiators, 1 test runner, with test client in different namespace. This allows the test runner to bypass the latency NetworkChaos, since it only affects the `from` direction

Also add a quick diagnostic check at the beginning of each NetworkLoadTest that hits `get_ledger_information`, which should reflect the cluster's configured network characteristics. Namely, that the latency between the Forge test runner and each of the validators should be low.

To test the network performance, use a bunch of local Forge canaries. In each, the following NetworkChaos are created and applied successfully. This can be done by observing the ChaosDashboard, checking the CRDs that exist in the namespace manually, as well as from experimental results (TPS and average client e2e latency)
```
$ kubectl get NetworkChaos
NAME                                  ACTION      DURATION
af-south-to-eu-north                  delay
af-south-to-us-west                   delay
eu-north-to-us-west                   delay
forge-namespace-100mbps-bandwidth     bandwidth
us-west-to-af-south                   delay
us-west-to-eu-north                   delay
```

Additionally, canary with Forge a bunch of times on this PR, peep the huge comment below (I've hidden it). After this, I've figured we should get rid of append-only sticky comment, and remove it.
The last run:
```
network::three-region-simulation : 2786 TPS, 736 ms latency, 73600 ms p99 latency,(!) expired 959563 out of 2163500 txns
Test Ok
```
[Grafana dashboard](https://o11y.aptosdev.com/grafana/d/overview/overview?orgId=1&refresh=10s&var-Datasource=Remote%20Prometheus%20Devinfra&var-namespace=forge-e2e-pr-4119&var-chain_name=forge-big-0&from=1663199827000&to=1663201034000)
[Humio Logs](https://cloud.us.humio.com/k8s/search?query=%24forgeLogs%28validator_instance%3D%2A%29%20%7C%20forge-e2e-pr-4119%20&live=false&start=1663199827000&end=1663201034000&widgetType=list-view&columns=%5B%7B%22type%22%3A%22field%22%2C%22fieldName%22%3A%22%40timestamp%22%2C%22format%22%3A%22timestamp%22%2C%22width%22%3A180%7D%2C%7B%22type%22%3A%22field%22%2C%22fieldName%22%3A%22level%22%2C%22format%22%3A%22text%22%2C%22width%22%3A54%7D%2C%7B%22type%22%3A%22link%22%2C%22openInNewBrowserTab%22%3A***%2C%22style%22%3A%22button%22%2C%22hrefTemplate%22%3A%22https%3A%2F%2Fgithub.com%2Faptos-labs%2Faptos-core%2Fpull%2F%7B%7Bfields%5B%5C%22github_pr%5C%22%5D%7D%7D%22%2C%22textTemplate%22%3A%22%7B%7Bfields%5B%5C%22github_pr%5C%22%5D%7D%7D%22%2C%22header%22%3A%22Forge%20PR%22%2C%22width%22%3A79%7D%2C%7B%22type%22%3A%22field%22%2C%22fieldName%22%3A%22k8s.namespace%22%2C%22format%22%3A%22text%22%2C%22width%22%3A104%7D%2C%7B%22type%22%3A%22field%22%2C%22fieldName%22%3A%22k8s.pod_name%22%2C%22format%22%3A%22text%22%2C%22width%22%3A126%7D%2C%7B%22type%22%3A%22field%22%2C%22fieldName%22%3A%22k8s.container_name%22%2C%22format%22%3A%22text%22%2C%22width%22%3A85%7D%2C%7B%22type%22%3A%22field%22%2C%22fieldName%22%3A%22message%22%2C%22format%22%3A%22text%22%7D%5D&newestAtBottom=***&showOnlyFirstLine=false)
[(Deprecated) OpenSearch Logs](https://es.devinfra.aptosdev.com/_dashboards/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'2022-09-14T23:57:07.000Z',to:'2022-09-15T00:17:14.000Z'))&_a=(columns:!(_source),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'d0bc5e20-badc-11ec-9a50-89b84ac337af',key:chain_name,negate:!f,params:(query:forge-big-0),type:phrase),query:(match_phrase:(chain_name:forge-big-0))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'d0bc5e20-badc-11ec-9a50-89b84ac337af',key:namespace,negate:!f,params:(query:forge-e2e-pr-4119),type:phrase),query:(match_phrase:(namespace:forge-e2e-pr-4119))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'d0bc5e20-badc-11ec-9a50-89b84ac337af',key:hostname,negate:!f,params:(query:aptos-node-0-validator-0),type:phrase),query:(match_phrase:(hostname:aptos-node-0-validator-0)))),index:'d0bc5e20-badc-11ec-9a50-89b84ac337af',interval:auto,query:(language:kuery,query:''),sort:!()))
[Test runner output](https://github.com/aptos-labs/aptos-core/actions/runs/3056711065)
Test run is land-blocking

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4119)

### CODE REVIEW

File: testsuite/forge.py [hunk 1]
```diff
@@ -1312,6 +1312,10 @@ def test(
     forge_cli_args: Optional[List[str]],
     test_args: Optional[List[str]],
 ) -> None:
+
+    # CANARY!!!
+    forge_test_suite = "network_latency"
```
Comment (rustielin) [inline testsuite/forge.py:1317]: added this to override the test suite ran by CI on PR.

File: testsuite/forge-cli/src/main.rs [hunk 2]
```diff
@@ -452,9 +452,13 @@ fn single_test_suite(test_name: &str) -> Result<ForgeConfig<'static>> {
             })),
         "config" => config.with_network_tests(vec![&ReconfigurationTest]),
         "network_partition" => config.with_network_tests(vec![&NetworkPartitionTest]),
-        "network_latency" => config
-            .with_network_tests(vec![&NetworkLatencyTest])
-            .with_success_criteria(SuccessCriteria::new(4000, 10000, true, None, None)),
+        "three_region_simulation" => config
+            .with_initial_validator_count(NonZeroUsize::new(20).unwrap())
+            .with_initial_fullnode_count(10)
```
Comment (sitalkedia) [inline testsuite/forge-cli/src/main.rs:457]: I don't think we need so many full nodes, 5 should be sufficient. I will take a look at other tests to optimize the number of FNs but let's try lower number for this test?

File: testsuite/forge-cli/src/main.rs [hunk 3]
```diff
@@ -452,9 +452,13 @@ fn single_test_suite(test_name: &str) -> Result<ForgeConfig<'static>> {
...
+            .with_emit_job(EmitJobRequest::default().mode(EmitJobMode::ConstTps { tps: 5000 }))
```
Comment (sitalkedia) [inline testsuite/forge-cli/src/main.rs:458]: Curious, did you try the mempool backlog load and what was the result from that?

File: testsuite/forge-cli/src/main.rs [hunk 4]
```diff
@@ -452,9 +452,13 @@ fn single_test_suite(test_name: &str) -> Result<ForgeConfig<'static>> {
...
+            .with_network_tests(vec![&ThreeRegionSimulationTest])
+            // TODO(rustielin): tune these success critiera after we have a better idea of the test behavior
+            .with_success_criteria(SuccessCriteria::new(1000, 100000, true, None, None)),
```
Comment (sitalkedia) [inline testsuite/forge-cli/src/main.rs:461]: Why so low TPS criteria? We can do more than 3k right?

File: testsuite/testcases/src/lib.rs [hunk 5]
```diff
@@ -212,6 +212,22 @@ impl dyn NetworkLoadTest {
...
             .map_err(|err| anyhow!("Failed to start runtime for transaction emitter. {}", err))?;
 
+        let clients = ctx
+            .swarm()
+            .get_clients_for_peers(&nodes_to_send_load_to, Duration::from_secs(10));
+
+        // Read first
+        for client in &clients {
+            let start = Instant::now();
+            let _v = rt.block_on(client.get_ledger_information())?;
```
Comment (sitalkedia) [inline testsuite/testcases/src/lib.rs:222]: Can you explain why is this needed?

File: testsuite/testcases/src/three_region_simulation_test.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+// 1 Gbps
+fn create_bandwidth_limit() -> SwarmNetworkBandwidth {
+    SwarmNetworkBandwidth {
+        rate: 1000,
+        limit: 20971520,
+        buffer: 10000,
+    }
+}
+
+// upper end of standard packet loss
```
Comment (sitalkedia) [inline testsuite/testcases/src/three_region_simulation_test.rs:102]: Please remove

File: testsuite/forge-cli/src/main.rs (hunk 3 above)
Comment (rustielin) [reply testsuite/forge-cli/src/main.rs:458]: Mempool backlog seemed to result in less reproducible results, which makes it hard for us to set a good SuccessCritiera TPS/latency too. I think the key is we just land this and start running it continuously, and tune from there.

File: testsuite/testcases/src/lib.rs (hunk 5 above)
Comment (rustielin) [reply testsuite/testcases/src/lib.rs:222]: This is just a sanity check that prints the latency of the connection from Forge test runner <---> validators.

File: testsuite/forge-cli/src/main.rs (hunk 4 above)
Comment (rustielin) [reply testsuite/forge-cli/src/main.rs]: bumped to 3k, but we can settle on a more reasonable number after we run it continuously and get a feel of the test result variance

File: testsuite/forge-cli/src/main.rs [hunk 7]
```diff
@@ -461,9 +461,13 @@ fn single_test_suite(test_name: &str) -> Result<ForgeConfig<'static>> {
                 Some(Duration::from_secs(240)),
                 None,
             )),
-        "network_latency" => config
-            .with_network_tests(vec![&NetworkLatencyTest])
-            .with_success_criteria(SuccessCriteria::new(4000, 10000, true, None, None)),
+        "three_region_simulation" => config
+            .with_initial_validator_count(NonZeroUsize::new(6).unwrap())
```
Comment (sitalkedia) [inline testsuite/forge-cli/src/main.rs:465]: I think we need few more validators to have enough number of validators per region? Let's make this 10 or 12?

**Review by sitalkedia — APPROVED** [review]

**Review by perryjrandall — APPROVED** [review]

File: testsuite/testcases/src/lib.rs [hunk 8]
```diff
@@ -212,6 +212,22 @@ impl dyn NetworkLoadTest {
...
+            let duration = start.elapsed();
+            info!(
+                "Fetch from {:?} took {}ms",
+                client.path_prefix_string(),
+                duration.as_millis(),
+            );
+        }
```
Comment (igor-aptos) [inline testsuite/testcases/src/lib.rs:229]: can we extract this code into a function so it is clear what it does?

Also serially calling all clients in this loop, can add a lot of latency (especially for larger 30-60 validator sets). let's make sure to issue the calls in parallel?
